### PR TITLE
expression: handle max_allowed_packet warnings for to_base64 function.

### PR DESCRIPTION
**Pull request by supernan1994**
## What have you changed? (mandatory)

Return NULL and a warning when the result of to_base64 exceeds max_allowed_packetbs.
Before this PR:
```
mysql> show variables like 'max_allowed_packet';
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| max_allowed_packet | 3    |
+--------------------+-------+
1 row in set (0.01 sec)

mysql> select to_base64("abc");
+------------------+
| to_base64("abc") |
+------------------+
| YWJj             |
+------------------+
1 row in set (0.00 sec)
```
After this PR:
```
mysql> show variables like 'max_allowed_packet';
+--------------------+-------+
| Variable_name      | Value |
+--------------------+-------+
| max_allowed_packet | 3    |
+--------------------+-------+
1 row in set (0.01 sec)

mysql> select to_base64("abc");
+------------------+
| to_base64("abc") |
+------------------+
| NULL             |
+------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+--------------------------------------------------------------------------+
| Level   | Code | Message                                                                  |
+---------+------+--------------------------------------------------------------------------+
| Warning | 1301 | Result of to_base64() was larger than max_allowed_packet (3) - truncated |
+---------+------+--------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

This PR calculates length of to_base64 result by function `base64NeededEncodedLength` before executing core of to_base64 logic, and compares length with global variable `max_allowed_packet`. 
unit tests are added for covering three circumstances:
1) result hasn't line feed
2) result has line feed, last few characters of result were filled with `=`
3) result has many line feed

## What is the type of the changes? (mandatory)
- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)
- unit test
- explain test
- manual test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
no

## Does this PR affect tidb-ansible update? (mandatory)
no

## Does this PR need to be added to the release notes? (mandatory)
Yes
Please note:
```
Return `NULL` when the result of function `TO_BASE64` is larger than `max_allowed_packet`
```

## Refer to a related PR or issue link (optional)
to #7153

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -3044,6 +3055,10 @@ func (b *builtinToBase64Sig) evalString(row chunk.Row) (d string, isNull bool, e
 		return "", isNull, errors.Trace(err)
 	}
 
+	if b.tp.Flen*mysql.MaxBytesOfCharacter > int(b.maxAllowedPacket) {
```
[the remark this answers is not in the document]
Comment (supernan1994) [reply expression/builtin_string.go]: sorry, thanks, I will correct it

[the remark this answers is not in the document]
Comment (zz-jason) [reply expression/builtin_string.go]: Hi, @supernan1994 any update?

[the remark this answers is not in the document]
Comment (zz-jason) [reply expression/builtin_string.go]: 😄 take your time, it's just a just a friendly Ping

File: expression/builtin_string_test.go [hunk 2]
```diff
@@ -1879,6 +1879,84 @@ func (s *testEvaluatorSuite) TestToBase64(c *C) {
...
+
+	for _, test := range tests {
+		resultType := &types.FieldType{Tp: mysql.TypeVarchar, Flen: base64NeededEncodedLength(len(test.args))}
+		base := baseBuiltinFunc{args: args, ctx: s.ctx, tp: resultType}
+		toBase64 := &builtinToBase64Sig{base, test.maxAllowPacket}
+
+		input := chunk.NewChunkWithCapacity(colTypes, 1)
+		input.AppendString(0, test.args)
+		res, isNull, err := toBase64.evalString(input.GetRow(0))
+		if test.getErr {
```
Comment (zz-jason) [inline expression/builtin_string_test.go]: all the added `test.getErr` is `false`, maybe this check can be replaced with: `c.Assert(err, IsNil)`

File: expression/builtin_string_test.go [hunk 3]
```diff
@@ -1879,6 +1879,84 @@ func (s *testEvaluatorSuite) TestToBase64(c *C) {
...
+		if test.isNil {
+			c.Assert(isNull, IsTrue)
+			warningCount += 1
+		} else {
+			c.Assert(isNull, IsFalse)
+		}
+		c.Assert(res, Equals, test.expect)
+	}
+	warnings := s.ctx.GetSessionVars().StmtCtx.GetWarnings()
+	c.Assert(len(warnings), Equals, warningCount)
```
Comment (zz-jason) [inline expression/builtin_string_test.go]: `warningCount` is always zero, seems the test of `max_allowed_packets` is not working

Comment (supernan1994) [reply expression/builtin_string_test.go]: warningCount will add 1 when `to_base64` result is nil

File: expression/builtin_string_test.go [hunk 4]
```diff
@@ -1879,6 +1879,75 @@ func (s *testEvaluatorSuite) TestToBase64(c *C) {
...
+		base := baseBuiltinFunc{args: args, ctx: s.ctx, tp: resultType}
+		toBase64 := &builtinToBase64Sig{base, test.maxAllowPacket}
+
+		input := chunk.NewChunkWithCapacity(colTypes, 1)
+		input.AppendString(0, test.args)
+		res, isNull, err := toBase64.evalString(input.GetRow(0))
+		c.Assert(err, IsNil)
+		if test.isNil {
+			c.Assert(isNull, IsTrue)
+			warningCount += 1
```
Comment (supernan1994) [inline expression/builtin_string_test.go]: `warningCount` will add 1 when `to_base64` result is nil

**Comment by zz-jason** [comment]
@supernan1994 re-triggered

File: expression/builtin_string_test.go (hunk 4 above)
Comment (zz-jason) [reply expression/builtin_string_test.go]: can we reset the warning appended to `s.ctx.GetSessionVars().StmtCtx` and check the exactly warning count and warning content in each test case?

Comment (supernan1994) [reply expression/builtin_string_test.go]: sure!

Comment (zz-jason) [reply expression/builtin_string_test.go]: 👍

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Comment by zz-jason** [comment]
@supernan1994 It would be nice if you can cherry-pick this PR to the `release-2.0` branch 😄

**Comment by supernan1994** [comment]
@zz-jason OK!

**Pull request merged**